tofhir-engine/src/main/scala/io/tofhir/engine/data/write/SparkArchiver.scala

```diff
@@ -0,0 +1,179 @@
+   * @param mappingErrors       data sources that got error because of the mapping definition
+   * @param invalidInputs       data sources that got error because of invalid input
+   */
+  def archiveDataSourcesOfFailedMappings(spark: SparkSession,
+                                         mappingJobExecution: FhirMappingJobExecution,
+                                         mappingUrl: Option[String],
+                                         notWrittenResources: util.List[FhirMappingResult],
+                                         mappingErrors: Dataset[FhirMappingResult],
+                                         invalidInputs: Dataset[FhirMappingResult]): Unit = {
+    if (mappingJobExecution.job.dataProcessingSettings.saveErroneousRecords) {
+      if (!invalidInputs.isEmpty) {
+        this.writeDatasetToConfiguredFolder(mappingJobExecution, invalidInputs, mappingUrl.get, FhirMappingErrorCodes.INVALID_INPUT)
+      }
+      if (!mappingErrors.isEmpty) {
+        this.writeDatasetToConfiguredFolder(mappingJobExecution, mappingErrors, mappingUrl.get, FhirMappingErrorCodes.MAPPING_ERROR)
+      }
+      if (!notWrittenResources.isEmpty) {
+        import spark.implicits._
+        val notWrittenResourcesDs = spark.createDataset[FhirMappingResult](notWrittenResources)
+        this.writeDatasetToConfiguredFolder(mappingJobExecution, notWrittenResourcesDs, mappingUrl.get, FhirMappingErrorCodes.INVALID_RESOURCE)
+      }
+    }
+  }
+
+  /**
+   * Writes the dataset to the configured folder
+   *
+   * @param mappingJobExecution job execution to get output directory of data sources with errors
+   * @param dataset             filtered dataset of data sources with errors to write to configured folder
+   * @param mappingUrl          to create directory for each mapping url within the job execution
+   * @param errorType           one of invalid_input, mapping_error, invalid_resource
+   */
+  private def writeDatasetToConfiguredFolder(mappingJobExecution: FhirMappingJobExecution,
+                                             dataset: Dataset[FhirMappingResult],
+                                             mappingUrl: String,
+                                             errorType: String): Unit = {
+    val outputPath = mappingJobExecution.getErrorOutputDirectory(mappingUrl, errorType)
+    val schema = schema_of_json(dataset.collect().head.source.get)
+
+    dataset
+      .withColumn("jsonData", from_json(col("source"), schema))
+      .select("jsonData.*")
+      .coalesce(1)
+      .write
+      .mode(SaveMode.Append)
+      .option("header", "true")
+      .csv(outputPath)
+
+    // Remove all files except the CSV file (to remove .crc files)
+    val srcFiles = FileUtil.listFiles(new File(outputPath))
+      .filterNot(f => f.getPath.endsWith(FileExtensions.CSV.toString))
+    srcFiles.foreach(f => f.delete())
+  }
+
+  /**
+   * Move file type data sources to archive folder
```

Comment:
I recommend to improve this description of the function since processArchiving is a braod term. It might be good to explain what is done in the match case conditions.